### ISSUE TITLE
Don't cast decimal literals implicitly to Float

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#70136]
Unmodified TPC-H Q6 gives wrong results (also see [here](https://clickhouse.com/docs/en/getting-started/example-datasets/tpch))

Original query, incorrect result:

```sql
SELECT
    sum(l_extendedprice * l_discount) AS revenue
FROM
    lineitem
WHERE
    l_shipdate >= DATE '1994-01-01'
    AND l_shipdate < DATE '1994-01-01' + INTERVAL '1' year
    AND l_discount BETWEEN 0.06 - 0.01 AND 0.06 + 0.01
    AND l_quantity < 24;
```

The problem is that decimal literals are immediately casted to `Float`.

Rewriten versions with correct result:

```sql
SELECT
    sum(l_extendedprice * l_discount) AS revenue
FROM
    lineitem
WHERE
    l_shipdate >= DATE '1994-01-01'
    AND l_shipdate < DATE '1994-01-01' + INTERVAL '1' year
    AND l_discount BETWEEN 0.05 AND 0.07
    AND l_quantity < 24;
```

(with explicit casts)

```sql
SELECT
    sum(l_extendedprice * l_discount) AS revenue
FROM
    lineitem
WHERE
    l_shipdate >= DATE '1994-01-01'
    AND l_shipdate < DATE '1994-01-01' + INTERVAL '1' year
    AND l_discount BETWEEN toDecimal32(0.06, 2) - toDecimal32(0.01, 2) AND toDecimal32(0.06,2) + toDecimal32(0.01, 2)
    AND l_quantity < 24;
```

**Comment by rienath** [comment]
As per standard, `<mantissa> E <exponent>` should be the only literal syntax that produces approximate numerics

**Comment by qoega** [comment]
We can't change this in ClickHouse and switch all to slower decimals. But we can delay literal parsing to a moment we have knowledge how we use it and applying minus to 2 numbers could be done in Decimal representation. When we have aggregation we can convert it to Floats without issues

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**Don't cast decimal literals implicitly to Float**

**ClickGap verified:** reproduced it with the script below on `master`; tested 6 version(s) — 6 affected, 0 clean (see Affects); long-standing (not a recent regression); introducing PR not yet pinned.

### Reproduction
```
-- A decimal-point literal is parsed as Float64, so 0.06 + 0.01 = 0.06999999999999999.
-- A Decimal column compared against that float silently drops the boundary row (0.07):
SELECT
    sumIf(price * discount, discount BETWEEN 0.06 - 0.01 AND 0.06 + 0.01) AS float_literals,
    sumIf(price * discount, discount BETWEEN toDecimal32(0.06,2) - toDecimal32(0.01,2) AND toDecimal32(0.06,2) + toDecimal32(0.01,2)) AS decimal_literals
FROM values('discount Decimal32(2), price Decimal32(2)', (0.05, 1000), (0.06, 1000), (0.07, 1000));
-- float_literals = 110  (0.07 row dropped),  decimal_literals = 180  (correct)

-- Root cause — the literal type:
SELECT toTypeName(0.06), 0.06 + 0.01;   -- Float64, 0.06999999999999999
```
[▶ Run on ClickHouse Fiddle](https://fiddle.clickhouse.com/9617c44e-efa7-4479-b667-516c1afdd144)

### Bisect
- Reproduces on every release line tested down to `v24.1.8.22` (the oldest with a downloadable binary for this repro). The introducing change predates `v24.1.8.22`, so it can't be pinned from release binaries — this is a **long-standing** issue, not a recent regression. A maintainer with full git history can `git blame` the relevant file/line to find the exact change.

### Affects
- **Reproduces on**: `master, 26.5, 26.4, 26.3, 26.2, 26.1`

### Suggested next step
Manual triage of the `comp-decimal` path is needed. The reproducer above should be a fast first check.

**Component:** `comp-decimal`
**Severity:** `P2`

---
_ClickGap · Finding: `h_issue_70136`_




### Triage metadata
- Primary component: `comp-decimal`
- Secondary components: `comp-query-analyzer`
- Possible duplicate: #70364 — Both report decimal literals being implicitly cast to Float in arithmetic causing precision loss; B explicitly references A as the same issue.
